Draft of a Working Group charter

grorg opened the pull request:
Initial draft of a charter for a Working Group. To be discussed at the 2019-01 meeting.

kainino0x commented:
[HTML version here](https://htmlpreview.github.io/?https://github.com/grorg/admin/blob/wg-charter-draft/wg-charter.html)

Kangz commented on wg-charter.html:
```diff
@@ -0,0 +1,364 @@
+        <ol>
+            <li>enables rendering of modern graphics to both onscreen and offscreen drawing surfaces</li>
+            <li>enables computation tasks to be performed, and the results of such tasks to be retrieved</li>
+            <li>defines a Shading Language that describes the graphics and computation tasks in a format
```
WHLSL is not part of the API itself and should not be tied to it since we have no consensus on using SPIRV or WHLSL as the ingestion format for WebGPU. All we agreed on is that it is a blessed language that could be defined in this group. (meaning we write some spec examples, tutorials and samples using it) 

I suggest "This group may also recommend a companion Shading Language that describes ..."

grorg replied:
✅

Kangz commented on wg-charter.html:
```diff
@@ -0,0 +1,364 @@
+              <p class="milestone"><b>Recommendation expected completion:</b> Q2 2020</p>
+          </dd>
+
+          <dt>WebGPU Shading Language (WHLSL)</dt>
```
This should be an optional deliverable, see comment above. Also I wouldn't callout the name WHLSL in this charter.

grorg replied:
✅

Kangz commented on wg-charter.html:
```diff
@@ -0,0 +1,364 @@
+            Normative Specifications
+          </h3>
+          <dl>
+          <dt>WebGPU API</dt>
```
nit: Web GPU API  :)

kainino0x replied:
Personally, I still would prefer to call it WebGPU 😛 (I like that it is a single word.)

grorg replied:
✅

Kangz commented on wg-charter.html:
```diff
@@ -0,0 +1,364 @@
+          Decision Policy
+        </h2>
+        <p>
+          This group will seek to make decisions through consensus and due process, per the <a href="https://www.w3.org/2018/Process-20180201/#Consensus"> W3C Process Document (section 3.3</a>). Typically, an editor or other participant makes an initial proposal, which is then refined in discussion with members of the group and other reviewers, and consensus emerges with little formal voting bein ...
```
I assume this is boilerplate that's also used for other groups but it is a bit weird to talk about proposals etc. given the WG would just stamp the specs.

grorg replied:
Yes, this is boilerplate.

gyagp commented on wg-charter.html:
```diff
@@ -0,0 +1,364 @@
+
+
+
+      <section id="scope" class="scope">
```
Usually there is a "Out of scope" part in this section to make the scope more clear. For example, we may not pursue a native program interface in this WG.

grorg replied:
✅

grorg commented:
I believe this is ready for the 2nd round of review. I looked at the WebAssembly charter and it doesn't say anything specific about the actual work happening in the Community Group. The draft here does, so we're being more explicit than the existing group that operates this way.

Kangz commented:
Thanks for the update, LGTM

kvark commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+
+        <p>
+            The API will not be restricted to any particular platform technology. Instead, it will
+            be generic enough to be implemented on top of modern GPU libraries, such as
```
We should change "libraries" to something like "system APIs"

anssiko commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+        <dt><a href="https://www.w3.org/community/wicg/">Web Platform Incubator Community Group</a></dt>
+        <dd>
+          This group provides a lightweight venue for proposing, incubating and discussing new web platform features. The Web Performance Working group will incubate and review new proposals that are within scope of our charter within the WICG. Once such WICG-incubated proposal is implemented and available in at least one major browser, and has support from one more, it may be adopted by the Web  ...
+        </dd>
```
@grorg, it may be useful to link back to https://webmachinelearning.github.io/charter/#coordination from here since participants of these two groups are already actively coordinating. Proposal:
```
<dt><a href="https://www.w3.org/community/webmachinelearning/">Machine Learning for the Web Community Group</a></dt>
<dd>This Community Group incubates a dedicated low-level Web API for machine learning inference.</dd>
```

grorg replied:
Good idea! Thanks @anssiko

mbergman42 commented:
The draft charter should define the acronym "GPU" and explain the context.  "GPU" is used in the draft in a way likely understandable to those in the industry.  But for someone not already in the discussion like myself, it was a bit of a speedbump to confirm that this was "graphics processing unit" and not some other "graphics" acronym with which I was as yet unfamiliar.  Still only "mostly sure"!

kainino0x commented:
Good catch. We can copy it from the bottom of the cg charter, which already has a definition.

grorg commented:
> Good catch. We can copy it from the bottom of the cg charter, which already has a definition.

Agreed!

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+            Group will be driving the technical work.
+        </p>
+
+        <div>
```
The success criteria section now appears in a separate section in the charter template, following discussion in: https://github.com/w3c/charter-drafts/issues/202
I suggest to move it a well.

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+
+        <div>
+          <h3>Success Criteria</h3>
+          <p>In order to advance to  <a href="https://www.w3.org/2018/Process-20180201/#RecsCR" title="Proposed Recommendation">Proposed Recommendation</a>, each specification is expected to have at least two independent implementations of each of feature defined in the specification.</p>
```
Please make "at least two independent implementations" a link to `https://www.w3.org/Consortium/Process/#implementation-experience`

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+              <p class="milestone"><b>Recommendation expected completion:</b> Q2 2020</p>
+          </dd>
+
+          <dt>WebGPU Test Suite</dt>
```
We usually don't list test suites as normative deliverables because they are essentially required to produce an implementation report and advance a specification to Proposed Recommendation.

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+        <div>
+          <h3>Success Criteria</h3>
+          <p>In order to advance to  <a href="https://www.w3.org/2018/Process-20180201/#RecsCR" title="Proposed Recommendation">Proposed Recommendation</a>, each specification is expected to have at least two independent implementations of each of feature defined in the specification.</p>
+          <p>The WebGPU specification should contain a section detailing any <a href='https://www.w3.org/TR/security-privacy-questionnaire/'>known security</a>, fingerprinting, and privacy implications, and suggested <a href='https://w3c.github.io/perf-security-privacy/'>mitigation strategies</a> for implementers, Web authors, and end users. The group should not publish a specification if accepta ...
```
Given that the group could work on another specification (WebGPU Shading Language in particular), I would s/The WebGPU specification/Each specification/ here and below.

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+          <dd>
+              An API as described by the <a
+              href="https://www.w3.org/community/gpu/">GPU for the Web
+              Community Group</a> charter.
```
There should be a `<p class="draft-status"><b>Draft state:</b> <a href="https://gpuweb.github.io/gpuweb/">Adopted from the GPU for the Web Community Group</a></p>` line here.

Also the expected completion date should be updated. Q4 2020 or Q2 2021, perhaps?

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+          <dt>WebGPU API</dt>
+
+          <dd>
+              An API as described by the <a
```
I would suggest "An API as described in the scope section" or copy the bits that you believe are relevant from the CG charter. The goal of a WG charter is to set things in stone, and the CG charter could potentially be updated outside of the usual WG rechartering process. This may be nitpicking though, what really matters is that the scope section does not change when external documents change.

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+        </p>
+        <p>
+          As stated above, the majority of the technical work for Working Group will take place in the <a href="https://www.w3.org/community/gpu/">GPU for the Web Community Group</a>.
+        </p>
```
Some more boilerplate text from the charter template to mention the CEPC:

```html
<p>
  Participants in the group are required
  (by the <a href="https://www.w3.org/Consortium/Process/#ParticipationCriteria">W3C Process</a>)
  to follow the W3C
  <a href="https://www.w3.org/Consortium/cepc/">Code of Ethics and Professional Conduct</a>.
</p>
```
```

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+            <td>
+              <strong>Teleconferences:</strong> as necessary.
+              <br />
+              <strong>Face-to-face:</strong> as necessary.
```
It is good practice to set expectations with regards to meetings. Typically, the CG has been meeting F2F regularly and has had weekly calls. If the expectation is that this will continue, then that should be specified here, e.g.: "Teleconferences: weekly or bi-weekly calls" and "Face-to-face: 3-4 times per year, including during W3C's annual Technical Plenary week"

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+        </ul>
+      </aside>
+      <p>
+        <a href="https://www.w3.org/"><img alt="W3C" height="48" src="http://www.w3.org/Icons/w3c_home" width="72" /></a>
```
https://www.w3.org/Icons/w3c_home
We're all HTTPS now ;)

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+        Dean Jackson
+      </address>
+
+      <p class="copyright">
```
Copyright year should be updated and a couple of HTTP links should move to HTTPS, correct copyright text:

```html
<p class="copyright">
  <a href="https://www.w3.org/Consortium/Legal/ipr-notice#Copyright">Copyright</a> ©
  2020
  <a href="https://www.w3.org/"><abbr title="World Wide Web Consortium">W3C</abbr></a><sup>®</sup>
  (
  <a href="https://www.csail.mit.edu/"><abbr title="Massachusetts Institute of Technology">MIT</abbr></a>,
  <a href="https://www.ercim.eu/"><abbr title="European Research Consortium for Informatics and Mathematics">ERCIM</abbr></a>,
  <a href="https://www.keio.ac.jp/">Keio</a>,
  <a href="https://ev.buaa.edu.cn/">Beihang</a>
  ), All Rights Reserved.

  <abbr title="World Wide Web Consortium">W3C</abbr> <a href="https://www.w3.org/Consortium/Legal/ipr-notice#Legal_Disclaimer">liability</a>, <a href="https://www.w3.org/Consortium/Legal/ipr-notice#W3C_Trademarks">trademark</a> and <a href="https://www.w3.org/Consortium/Legal/copyright-documents">document use</a> rules apply.
</p>
```

grorg commented:
I will do one more editing pass asap.

After that, we've decided to use the reviewing approval as an indication that your company (and possibly its legal team) have reviewed the charter and consider it ok to move forward.

We'll be checking in on this every week.

Kangz commented:
Thanks for the detailed comments!

@grorg do you have some cycles to update this PR?

tidoust commented:
@grorg, @Kangz, would you want me to do the boring work and integrate remaining comments? I'd like to start horizontal (accessibility, internationalizaton, privacy, security) reviews on the draft charter.

grorg commented:
@tidoust appreciated! please go for it!

tidoust commented:
OK, @grorg I did that as a [pull request against your fork](https://github.com/grorg/admin/pull/1) so that the update appears in this thread once merged there. I can also create a separate PR in this repo if that approach feels too convoluted...

A couple of questions for the group:
1. Now that the group seems to have converged on a solution for the Shading Language, should the draft charter be more affirmative there and make WGSL a proper deliverable (instead of a possible one)?
2. I put Q3 2021 as a completion date for the WebGPU API. It's always hard to evaluate how long things are going to take. You have more visibility than I have on what needs to be done, and on how long it could take, so how does that look? Q3 2021 could be a bit optimistic for instance.

wseltzer commented:
Hi folks! I know things are busy for everyone. Is there anything more W3C team can do to help get a canonical version of this draft charter so we can circulate for wider review?

grorg commented:
I've merged @tidoust 's PR on my PR, which addresses the comments above. Thanks @tidoust .

@wseltzer I think we have everything we need from W3C.

At this point I think the reviewers (i.e. member companies) need to do another pass to sign off.

Kangz commented on wg-charter.html:
```diff
@@ -0,0 +1,387 @@
+        <div>
+          <h3>Success Criteria</h3>
+          <p>In order to advance to  <a href="https://www.w3.org/2018/Process-20180201/#RecsCR" title="Proposed Recommendation">Proposed Recommendation</a>, each specification is expected to have at least two independent implementations of each of feature defined in the specification.</p>
+          <p>The WebGPU specification should contain a section detailing any <a href='https://www.w3.org/TR/security-privacy-questionnaire/'>known security</a>, fingerprinting, and privacy implications, and suggested <a href='https://w3c.github.io/perf-security-privacy/'>mitigation strategies</a> for implementers, Web authors, and end users. The group should not publish a specification if accepta ...
```
I think this comment is still outstanding, but it seems to me that this text is more boilerplate for all W3C specifications than specific to WebGPU itself.

"should" gives options on how to organize such sections, while giving a clear goal to address these topics.

tidoust replied:
Right, I stuck to the usual boilerplate text. The need to address (and seek review for) security, fingerprinting, privacy and other horizontals is a must in any case. I personally think it's ok to give some leeway for the group on ways to crystallizes the outcomes in the spec, but the text can be strengthened out if people prefer.

Kangz commented on wg-charter.html:
```diff
@@ -0,0 +1,406 @@
+          <h3 id="w3c-coordination">W3C Groups</h3>
+          <dl>
+            <dt><a href="https://www.w3.org/2001/tag/">Technical Architecture Group (TAG)</a></dt>
+        <dd>The WG may ask the Technical Architecture Group to review some of its specifications.</dd>
```
This should be "will ask" instead of "may ask" to match what's said in 'coordination":

> For all specifications, this Working Group will seek horizontal review ... , and with the TAG

tidoust replied:
That's a good point, but I believe the fix is to drop the TAG from this list, because there's simply no reason to include it in the first place. That list is meant to convey "additional" groups with which the GPU for the Web WG intends to exchange on top of those that will de facto run reviews on the specs such as the TAG.

tidoust commented:
I prepared another pull request (see https://github.com/grorg/admin/pull/2) to address feedback received so far here and in https://github.com/w3c/strategy/issues/68.

Main changes:
- Firm up stance on WebGPU Shading Language: move spec to list of deliverables, mention SPIR-V and key goals.
- Mention mitigation strategies for fingerprinting and unauthorized use of computational resources in Scope (on top of in success criteria section). Or do you already have a clear idea of ways to address these issues?

More minor changes:
- Drop mentions of Web Performance Working Group and link to Web Perf doc
- Drop TAG, WICG and TC 39 groups from coordination list.
- Add Immersive Web Working Group to coordination list as a likely consumer of the WebGPU API.
- Update boilerplate text for Coordination section to match current template.
- Postpone completion milestones to dates that seem more realistic.

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,406 @@
+          Participation
+        </h2>
+        <p>
+          To be successful, this Working Group is expected to have 5 or more active participants for its duration, including representatives from the key implementors of this specification, and active Editors and Test Leads for each specification. The Chairs, specification Editors, and Test Leads are expected to contribute half of a day per week towards the Working Group. There is no minimum requ ...
```
Some groups choose to nominate test leads to lead the development of the test suites. That is not as formal a role as editor, and not a mandatory role either. The (boilerplate) text is more meant to convey the fact that work on test suites is part of success criteria for a group.

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,406 @@
+        <dt><a href="https://www.w3.org/community/webmachinelearning/">Machine Learning for the Web Community Group</a></dt>
+        <dd>This Community Group incubates a dedicated low-level Web API for machine learning inference.</dd>
+
+        <dt><a href="https://www.w3.org/community/wicg/">Web Platform Incubator Community Group</a></dt>
```
I dropped the WICG. I suspect this was the result of a copy-and-paste from the Web Performance WG charter...

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,406 @@
+          </dl>
+          <h3 id="external-coordination">External Organizations</h3>
+          <dl>
+            <dt><a href="https://www.ecma-international.org/memento/tc39.htm" id="tc39">ECMA Technical Committee 39 (TC39)</a></dt>
```
Same here. I dropped that in latest PR.

tidoust commented on wg-charter.html:
```diff
@@ -0,0 +1,406 @@
+              A Shading Language specification that defines the programmable
+              interface to the GPU for the Web graphics and computation
+              pipeline. Alternatively, the Working Group may recommend
+              an existing Shading Language.
```
Note the PR I prepared firms things up, moving the spec to the list of deliverables... or is too firm now?

Kangz commented:
LGTM @grorg can you take a look at the PR?

tidoust commented:
@grorg Are you still looking at the proposed update? Could it be merged?

tidoust commented:
Unless I hear pushback, I note that I'm planning to move to next stage this week (internal review by W3C Management, which should lead to the draft charter being sent to the Advisory Committee for official review)

tidoust commented:
W3C issued a call for review of the proposed charter to its Advisory Committee today. Public version at:
https://lists.w3.org/Archives/Public/public-new-work/2020Jun/0008.html

I made a copy of the draft charter for the review, hosted on w3.org:
https://www.w3.org/2020/06/gpuweb-charter.html

This copy integrates changes that the CG already reviewed and a couple of updates requested during internal review (notably to add the GPU for the Web CG appears under Coordination since it wasn't there). Updates since the version the CG reviewed:
- https://github.com/grorg/admin/pull/2/commits/20e9413bb1e115ce95565055661b002fd66e85b9
- https://github.com/grorg/admin/pull/2/commits/8a9b7bd661c94bd1e7e637ba57aa2d280cd6cd46

(All the updates are in https://github.com/grorg/admin/pull/2)

Call for review lasts 6 weeks until 31 July 2020. Please work with your Advisory Committee Representative to register your comments and support for the charter.

tidoust commented:
Closing this outdated pull request. Charter was adopted two years ago. Published version at https://www.w3.org/2020/12/gpu-wg-charter.html
